x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/PasswordBasedEncryptor.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.repositories.encrypted;
+
+import org.apache.logging.log4j.util.Supplier;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.collect.Tuple;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Encrypts and decrypts data using a password.
+ * Encryption generates the AES secret key from the password and a randomly generated salt (using the PBKDF2 algo).
+ * This key is then used to encrypt the data (AES/GCM/NoPadding). The encryption IV is generated randomly.
+ * The key is cached internally and used for at most {@link #ENCRYPT_INVOKE_LIMIT_USING_SAME_KEY} encryption invocations.
+ * When this limit is exceeded, a new key is generated, by generating a new random salt. The encryption key is never
+ * stored on disk. The salt, however, which was used in generating the encryption key, is prepended to the ciphertext.
+ * Decryption extracts the salt prepended to the ciphertext, computes the key (using the secret password) and uses
+ * the key to decrypt the ciphertext (which also contains the IV). Decryption also does not store the generated key
+ * on disk, but caches it in memory because generating the key from the password is computationally expensive on purpose.
+ */
+public final class PasswordBasedEncryptor {
```

Comment:
@tvernum @jkakavas Can you also please mull over this as well? This describes how metadata encryption works. It is different compared to how blobs are encrypted because for the metadata, compared to the actual blob, we should reuse the key (which is derived from the password); but if we reuse the key we then have to carefully plan the construction of IVs. In https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf Section 8.2 it is recommended to recycle the key after a certain number of randomly generated IVs . To generate multiple keys from the same password I use here a random salt. The salt is then prepended to the ciphertext for the decryption to be able to regenerate the key. The key is never stored on disk (so it's not stored near the salt).

I believe this is secure. Would you agree?

This replaces CMS, because of its complicated API and algorithm limitations.

Reply:
I agree that the proposed implementation adheres to the recommendations and doesn't violate any of the limitations that would render its use insecure.

Have you consider the possibility/performance of the naive approach of generating a new salt for each encryption operation ? We can be certain that this is less performant, but I can't necessarily gauge _how much_ and whether or not it would make sense as a trade-off in order to make the implementation less complex.

Reply:
Thanks a lot for verifying it!

I haven't benchmarked it.
I think it's hard to say what performance penalty would be acceptable, because it must be compared with the latency of the "put/get-blob" operation, which has a lot of variability and also the PBKDF2 function itself, I suspect, has a lot of variability on different architectures. Basically, I worry that the results will be inconclusive.

I wonder what @tvernum thinks about the complexity here?

In the mean time, I will create tests for this class to see if we can cover for the complexity.

@jkakavas Do you think I should maybe fork this class, and its tests, into its own PR?